mesonbuild/cargo/interpreter.py

```diff
@@ -0,0 +1,416 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright © 2022-2023 Intel Corporation
+
+"""Interpreter for converting Cargo Toml definitions to Meson AST
+
+There are some notable limits here. We don't even try to convert something with
+a build.rs, there's so few limits on what Cargo alows a build.rs (basically
+none), and no good way for us to convert time. In that case, an actual meson port
+will be required.
+"""
+
+from __future__ import annotations
+import dataclasses
+import glob
+import importlib
+import itertools
+import os
+import typing as T
+
+from . import builder
+from . import version
+from .. import mparser
+from .._pathlib import Path
+
+if T.TYPE_CHECKING:
+    from types import ModuleType
+
+    from . import manifest
+    from ..environment import Environment
+
+# tomllib is present in python 3.11, before that it is a pypi module called tomli,
+# we try to import tomllib, then tomli,
+# TODO: add a fallback to toml2json?
```

Comment:
I would go ahead and add the fallback path, and also add a custom error if tomllib is still none and we can't find the fallback executable.

Reply:
IIRC in the original PR it checks if the module is none when doing a cargo subproject, and it's not fatal if required=false.

Reply:
Yeah, my goal was originally that we'd basically do a check in the cargo subproject call that fails if required == true. but implementing a path for toml2json before landing this probably make sense

Reply:
Ah okay I see, the part that catches the error is not included in this PR because it's been pared down to the mesonbuild/cargo/ directory only and that was part of mesonbuild/modules/.

Reply:
What value does it add to have toml2json? One dependency or another does not make much difference, it just have to be installed with pip or as distro dependency. better not have too many code paths IMHO.

Reply:
"installed with pip" would be the main thing I think. As a command-line tool there's a lot of ways to install it -- I know of at least 5 projects that all create an executable named `toml2json`, so you can take your pick of C++, rust, golang, or even python again. But the C++ one by @dcbaker is trivially bootstrapped.

Reply:
I did not realize that's a CLI option. I guess that could be added, but I don't think that should be blocker to the initial merge.